x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -394,6 +394,32 @@ void getAndValidateToken(ThreadContext ctx, ActionListener<UserToken> listener)
         }
     }
 
+    /**
+     * Decodes the provided token, and validates it (for format, expiry and invalidation).
+     * If valid, the token's {@link Authentication} (see {@link UserToken#getAuthentication()} is provided to the listener.
+     * If the token is invalid (expired etc), then {@link ActionListener#onFailure(Exception)} will be called.
+     * If tokens are not enabled, or the token does not exist, {@link ActionListener#onResponse} will be called with a
+     * {@code null} authentication object.
+     */
+    public void authenticateToken(SecureString tokenString, ActionListener<Authentication> listener) {
+        if (isEnabled()) {
+            decodeToken(tokenString.toString(), ActionListener.wrap(userToken -> {
+                if (userToken != null) {
+                    checkIfTokenIsValid(userToken, ActionListener.wrap(
+                        token -> {
+                            listener.onResponse(token == null ? null : token.getAuthentication());
+                        },
+                        listener::onFailure
+                    ));
+                } else {
+                    listener.onResponse(null);
```

Comment:
Also because this new method is not used in the authn chain, when a token fails decoding it's a client error.
I believe:
```
listener.onFailure(new IllegalArgumentException());
```
would do, it would return 400 (currently this falls back to 500 in `ApiKeyGenerator#generateApiKey`).

Reply:
Good point @albertzaharovits